source/Cosmos.IL2CPU/IL/Conv_I2.cs

```diff
@@ -24,48 +21,45 @@ public override void Execute(_MethodInfo aMethod, ILOpCode aOpCode)
             var xSource = aOpCode.StackPopTypes[0];
             var xSourceSize = SizeOfType(xSource);
             var xSourceIsFloat = TypeIsFloat(xSource);
-            if (xSourceIsFloat)
+
+            if (xSourceSize > 8)
+            {
+                throw new NotImplementedException("Cosmos.IL2CPU.x86->IL->Conv_I2.cs->Error: StackSize > 8 not supported");
+            }
+
+            if (xSourceSize <= 4)
             {
-                if (xSourceSize == 4)
+                if (xSourceIsFloat)
                 {
                     XS.SSE.MoveSS(XMM0, ESP, sourceIsIndirect: true);
                     XS.SSE.ConvertSS2SIAndTruncate(EAX, XMM0);
+                    XS.MoveSignExtend(EAX, AX);
                     XS.Set(ESP, EAX, destinationIsIndirect: true);
                 }
-                else if (xSourceSize == 8)
+                else
+                {
+                    XS.Pop(EAX);
+                    XS.MoveSignExtend(EAX, AX);
+                    XS.Push(EAX);
+                }
+            }
+            else if (xSourceSize <= 8)
+            {
+                if (xSourceIsFloat)
                 {
                     XS.SSE2.MoveSD(XMM0, ESP, sourceIsIndirect: true);
+                    XS.Add(ESP, 4);
                     XS.SSE2.ConvertSD2SIAndTruncate(EAX, XMM0);
                     XS.Set(ESP, EAX, destinationIsIndirect: true);
                 }

```

Comment:
The problem is that stack size 1, 2 and 4 are different, as they need to be handled differently in load instructions (stack sizes 1 and 2 need to be extended), so I can't think of a simple way of matching "stack size 1, 2 or 4 and not float".